### DISCUSSION TITLE
Implicit string cast when attempting to add int to string

### DISCUSSION THREAD

**Discussion by DukeOfSuffix** [dotnet/csharplang#5667] [Q&A] [answered]
I was demonstrating static typing to a friend who is trying to learn C#. When I tried to add a string to an integer, it seems to implicitly cast the int to a string and perform a string concatenation:

![image](https://user-images.githubusercontent.com/97639190/150005614-0351d4bd-5432-4ac2-805c-e4e9624c5d07.png)

Was this added to C# recently, or is it a bug?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Was this added to C# recently,

Yes.  Though 'recent' only relative to the age of the universe.  This functionality was available in C#1.0 

>  or is it a bug?

It is not a bug.  (Practically) all types in .Net extend from `System.Object` and thus have the ability to convert themselves to strings with `.ToString()`.  String concatenation is thus supported in C# for all of these types through the use of combining the actual strings, and the stringified values into one final string.

> I was demonstrating static typing to a friend who is trying to learn C#

Note: C# is statically typed.  And in this scenario static typing is in effect.  Static typing does not mean that strings cannot be concatenated to non-string values.

**Reply by DukeOfSuffix** [reply]
Regarding your final point, I didn't mean to imply that C# *wasn't* statically typed. I just thought that the behavior may not be consistent with the spirit of static typing (i.e. the integration of two fundamentally different data types).

**Reply by CyrusNajmabadi** [reply]
>  (i.e. the integration of two fundamentally different data types).

Note: static typing has no problem with this :)

Static typing just means that the types are known and checked at compile time *prior* to running your program.  Here the C# compiler knew the types here and checked that what you were doing was ok.  It determined that it's 100% ok and it allowed this code and compiled it into something sensible :)

**Comment by mburbea** [comment]
The `+` operator in C# when applied to string as an operand is sugar for `string.Concat`. The compiler will call `ToString` on all non-strings in the invocation, and the appropriate overload of string.Concat will be called. See this sharplab:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGAzAATakDCpA3oaQ+WRjgAykCyAFAJS32NBGAOykARJADuYgNQ4sAbgEMAvoRVA
